File: ao-processes/proof-corpus.lua
local json = require('json')

ADMIN = ADMIN or Owner

State = State or {
  version = 1,
  chain = "ethereum",
  genesis = nil,
  latest = nil,
  segments = {},
  facts = {},
  pending_facts = 0,
  rejected_count = 0
}

local function isAdmin(msg)
  return ADMIN ~= nil and msg.From == ADMIN
end

local function reply(target, action, data)
  ao.send({
    Target = target,
    Action = action,
    Data = json.encode(data)
  })
end

local function decode(data)
  local ok, value = pcall(json.decode, data or "{}")
  if not ok or type(value) ~= "table" then
    return nil, "Invalid JSON"
  end
  return value, nil
end

local function isHexRoot(value)
  return type(value) == "string" and string.match(value, "^0x[%da-fA-F]+$") and string.len(value) == 66
end

local function requireField(tbl, field)
  if tbl[field] == nil or tbl[field] == "" then
    return field .. " is required"
  end
  return nil
end

local function countTable(tbl)
  local count = 0
  for _ in pairs(tbl) do
    count = count + 1
  end
  return count
end

Handlers.add(
  "initialize-genesis",
  function(msg)
    return msg.Action == "InitializeGenesis"
  end,
  function(msg)
    if not isAdmin(msg) then
      reply(msg.From, "GenesisRejected", { error = "Sender is not admin" })
      return
    end
    if State.genesis ~= nil then
      reply(msg.From, "GenesisRejected", { error = "Genesis already initialized" })
      return
    end

    local data, err = decode(msg.Data)
    if err then
      reply(msg.From, "GenesisRejected", { error = err })
      return
    end
    if data.chain ~= nil and data.chain ~= "ethereum" then
      reply(msg.From, "GenesisRejected", { error = "Unsupported chain" })
      return
    end
    if not isHexRoot(data.state_root) then
      reply(msg.From, "GenesisRejected", { error = "state_root must be a 32-byte hex root" })
      return
    end
    if type(data.commitment) ~= "string" or data.commitment == "" then
      reply(msg.From, "GenesisRejected", { error = "commitment is required" })
      return
    end

    State.genesis = {
      chain = "ethereum",
      slot = 0,
      state_root = data.state_root,
      commitment = data.commitment
    }
    State.latest = {
      slot = 0,
      state_root = data.state_root,
      commitment = data.commitment
    }

    reply(msg.From, "GenesisInitialized", State.genesis)
  end
)

Handlers.add(
  "submit-segment-proof",
  function(msg)
    return msg.Action == "SubmitSegmentProof"
  end,
  function(msg)
    if State.genesis == nil or State.latest == nil then
      reply(msg.From, "SegmentRejected", { error = "Genesis not initialized" })
      return
    end

    local data, err = decode(msg.Data)
    if err then
      reply(msg.From, "SegmentRejected", { error = err })
      return
    end

    for _, field in ipairs({ "from_slot", "to_slot", "from_state_root", "to_state_root", "proof_system", "proof_hash", "commitment" }) do
      local fieldErr = requireField(data, field)
      if fieldErr then
        State.rejected_count = State.rejected_count + 1
        reply(msg.From, "SegmentRejected", { error = fieldErr })
        return
      end
    end

    if tonumber(data.from_slot) ~= tonumber(State.latest.slot) then
      State.rejected_count = State.rejected_count + 1
      reply(msg.From, "SegmentRejected", { error = "Segment must start at latest slot" })
      return
    end
    if tonumber(data.to_slot) <= tonumber(data.from_slot) then
      State.rejected_count = State.rejected_count + 1
      reply(msg.From, "SegmentRejected", { error = "to_slot must be greater than from_slot" })
      return
    end
    if data.from_state_root ~= State.latest.state_root then
      State.rejected_count = State.rejected_count + 1
      reply(msg.From, "SegmentRejected", { error = "Segment must start at latest state root" })
      return
    end
    if not isHexRoot(data.from_state_root) or not isHexRoot(data.to_state_root) then
      State.rejected_count = State.rejected_count + 1
      reply(msg.From, "SegmentRejected", { error = "State roots must be 32-byte hex roots" })
      return
    end

    local segment = {
      from_slot = tonumber(data.from_slot),
      to_slot = tonumber(data.to_slot),
      from_state_root = data.from_state_root,
      to_state_root = data.to_state_root,
      proof_system = data.proof_system,
      proof_hash = data.proof_hash,
      commitment = data.commitment,
      submitted_by = msg.From,
      timestamp = os.time()
    }

    table.insert(State.segments, segment)
    State.latest = {
      slot = segment.to_slot,
      state_root = segment.to_state_root,
      commitment = segment.commitment
    }

    reply(msg.From, "SegmentAccepted", {
      latest_slot = State.latest.slot,
      latest_state_root = State.latest.state_root,
      latest_commitment = State.latest.commitment,
      segment_count = #State.segments
    })
  end
)

Handlers.add(
  "submit-fact-proof",
  function(msg)
    return msg.Action == "SubmitFactProof"
  end,
  function(msg)
    if State.latest == nil then
      reply(msg.From, "FactRejected", { error = "Corpus not initialized" })
      return
    end

    local data, err = decode(msg.Data)
    if err then
      reply(msg.From, "FactRejected", { error = err })
      return
    end

    for _, field in ipairs({ "slot", "state_root", "predicate", "subject", "value", "corpus_commitment", "proof_hash", "commitment" }) do
      local fieldErr = requireField(data, field)
      if fieldErr then
        State.rejected_count = State.rejected_count + 1
        reply(msg.From, "FactRejected", { error = fieldErr })
        return
      end
    end

    if data.corpus_commitment ~= State.latest.commitment then
      State.rejected_count = State.rejected_count + 1
      reply(msg.From, "FactRejected", { error = "Fact must bind to latest corpus commitment" })
      return
    end
    if tonumber(data.slot) > tonumber(State.latest.slot) then
      State.rejected_count = State.rejected_count + 1
      reply(msg.From, "FactRejected", { error = "Fact slot exceeds latest proven slot" })
      return
    end
    if not isHexRoot(data.state_root) then
      State.rejected_count = State.rejected_count + 1
      reply(msg.From, "FactRejected", { error = "state_root must be a 32-byte hex root" })
      return
    end

    local key = tostring(data.slot) .. ":" .. data.predicate .. ":" .. data.subject
    State.facts[key] = {
      slot = tonumber(data.slot),
      state_root = data.state_root,
      predicate = data.predicate,
      subject = data.subject,
      value = data.value,
      corpus_commitment = data.corpus_commitment,
      proof_hash = data.proof_hash,
      commitment = data.commitment,
      submitted_by = msg.From,
      timestamp = os.time()
    }

    reply(msg.From, "FactAccepted", {
      key = key,
      commitment = data.commitment,
      latest_slot = State.latest.slot
    })
  end
)

Handlers.add(
  "get-corpus-state",
  function(msg)
    return msg.Action == "GetCorpusState"
  end,
  function(msg)
    reply(msg.From, "CorpusState", {
      version = State.version,
      chain = State.chain,
      genesis = State.genesis,
      latest = State.latest,
      segment_count = #State.segments,
      fact_count = countTable(State.facts),
      rejected_count = State.rejected_count
    })
  end
)

Handlers.add(
  "get-fact",
  function(msg)
    return msg.Action == "GetFact"
  end,
  function(msg)
    local data, err = decode(msg.Data)
    if err then
      reply(msg.From, "FactResult", { error = err })
      return
    end
    local key = tostring(data.slot) .. ":" .. tostring(data.predicate) .. ":" .. tostring(data.subject)
    reply(msg.From, "FactResult", State.facts[key] or { error = "not found", key = key })
  end
)

File: core/proof-corpus.js
const { hashState } = require("./hash");

const ETHEREUM_GENESIS_ROOT =
  "0xd7f8974fb5ac78d9ac099b9b430502e5ccefe0f1d5518b5fdb5c23b3183a6823";

function assert(condition, message) {
  if (!condition) {
    throw new Error(message);
  }
}

function isHexRoot(value) {
  return typeof value === "string" && /^0x[0-9a-fA-F]{64}$/.test(value);
}

function hashCommitment(kind, payload) {
  return hashState({
    paxiom_commitment_kind: kind,
    payload
  });
}

function createGenesisAnchor(options = {}) {
  const chain = options.chain || "ethereum";
  const genesisRoot = options.genesis_root || ETHEREUM_GENESIS_ROOT;

  assert(chain === "ethereum", `Unsupported chain: ${chain}`);
  assert(isHexRoot(genesisRoot), "Genesis root must be a 32-byte hex root");

  return {
    kind: "GENESIS_ANCHOR",
    chain,
    slot: 0,
    block_number: 0,
    state_root: genesisRoot,
    commitment: hashCommitment("GENESIS_ANCHOR", {
      chain,
      slot: 0,
      block_number: 0,
      state_root: genesisRoot
    })
  };
}

function createSegmentProof(input) {
  assert(input && typeof input === "object", "Segment proof input must be an object");

  const {
    chain = "ethereum",
    from_slot,
    to_slot,
    from_state_root,
    to_state_root,
    proof_system,
    proof_hash,
    public_inputs = {},
    dependencies = []
  } = input;

  assert(chain === "ethereum", `Unsupported chain: ${chain}`);
  assert(Number.isInteger(from_slot) && from_slot >= 0, "from_slot must be a non-negative integer");
  assert(Number.isInteger(to_slot) && to_slot > from_slot, "to_slot must be greater than from_slot");
  assert(isHexRoot(from_state_root), "from_state_root must be a 32-byte hex root");
  assert(isHexRoot(to_state_root), "to_state_root must be a 32-byte hex root");
  assert(typeof proof_system === "string" && proof_system.length > 0, "proof_system is required");
  assert(typeof proof_hash === "string" && proof_hash.length > 0, "proof_hash is required");
  assert(Array.isArray(dependencies), "dependencies must be an array");

  const segment = {
    kind: "SEGMENT_PROOF",
    chain,
    from_slot,
    to_slot,
    from_state_root,
    to_state_root,
    proof_system,
    proof_hash,
    public_inputs,
    dependencies
  };

  return {
    ...segment,
    commitment: hashCommitment("SEGMENT_PROOF", segment)
  };
}

function verifyContiguousSegments(segments) {
  assert(Array.isArray(segments) && segments.length > 0, "At least one segment is required");

  for (let i = 0; i < segments.length; i++) {
    const segment = segments[i];
    assert(segment.kind === "SEGMENT_PROOF", `Invalid segment kind at index ${i}`);
    assert(segment.commitment === hashCommitment("SEGMENT_PROOF", {
      kind: segment.kind,
      chain: segment.chain,
      from_slot: segment.from_slot,
      to_slot: segment.to_slot,
      from_state_root: segment.from_state_root,
      to_state_root: segment.to_state_root,
      proof_system: segment.proof_system,
      proof_hash: segment.proof_hash,
      public_inputs: segment.public_inputs,
      dependencies: segment.dependencies
    }), `Segment commitment mismatch at index ${i}`);

    if (i > 0) {
      const previous = segments[i - 1];
      assert(segment.from_slot === previous.to_slot, `Slot gap before segment ${i}`);
      assert(segment.from_state_root === previous.to_state_root, `State root gap before segment ${i}`);
    }
  }
}

function aggregateSegmentProofs(segments, options = {}) {
  verifyContiguousSegments(segments);

  const first = segments[0];
  const last = segments[segments.length - 1];
  const aggregate = {
    kind: "AGGREGATE_PROOF",
    chain: first.chain,
    from_slot: first.from_slot,
    to_slot: last.to_slot,
    from_state_root: first.from_state_root,
    to_state_root: last.to_state_root,
    proof_system: options.proof_system || "recursive-placeholder",
    aggregation_strategy: options.aggregation_strategy || "ordered-contiguous-fold",
    segment_commitments: segments.map(segment => segment.commitment),
    public_inputs: options.public_inputs || {}
  };

  return {
    ...aggregate,
    commitment: hashCommitment("AGGREGATE_PROOF", aggregate)
  };
}

function createFactProof(input) {
  assert(input && typeof input === "object", "Fact proof input must be an object");

  const {
    chain = "ethereum",
    slot,
    state_root,
    predicate,
    subject,
    value,
    corpus_commitment,
    witness_hash,
    proof_hash,
    proof_system
  } = input;

  assert(chain === "ethereum", `Unsupported chain: ${chain}`);
  assert(Number.isInteger(slot) && slot >= 0, "slot must be a non-negative integer");
  assert(isHexRoot(state_root), "state_root must be a 32-byte hex root");
  assert(typeof predicate === "string" && predicate.length > 0, "predicate is required");
  assert(typeof subject === "string" && subject.length > 0, "subject is required");
  assert(typeof corpus_commitment === "string" && corpus_commitment.length > 0, "corpus_commitment is required");
  assert(typeof witness_hash === "string" && witness_hash.length > 0, "witness_hash is required");
  assert(typeof proof_hash === "string" && proof_hash.length > 0, "proof_hash is required");
  assert(typeof proof_system === "string" && proof_system.length > 0, "proof_system is required");

  const fact = {
    kind: "FACT_PROOF",
    chain,
    slot,
    state_root,
    predicate,
    subject,
    value,
    corpus_commitment,
    witness_hash,
    proof_hash,
    proof_system
  };

  return {
    ...fact,
    commitment: hashCommitment("FACT_PROOF", fact)
  };
}

function createProofCorpus(genesis = createGenesisAnchor()) {
  return {
    version: 1,
    chain: genesis.chain,
    genesis,
    latest_aggregate: null,
    segment_commitments: [],
    fact_commitments: [],
    index: {}
  };
}

function attachAggregate(corpus, aggregate) {
  assert(corpus && corpus.version === 1, "Invalid proof corpus");
  assert(aggregate && aggregate.kind === "AGGREGATE_PROOF", "Invalid aggregate proof");
  assert(aggregate.chain === corpus.chain, "Aggregate chain mismatch");
  assert(aggregate.from_slot === corpus.genesis.slot, "Aggregate must start at genesis slot");
  assert(aggregate.from_state_root === corpus.genesis.state_root, "Aggregate must start at genesis root");

  return {
    ...corpus,
    latest_aggregate: aggregate,
    segment_commitments: aggregate.segment_commitments.slice(),
    index: {
      ...corpus.index,
      latest_slot: aggregate.to_slot,
      latest_state_root: aggregate.to_state_root,
      latest_commitment: aggregate.commitment
    }
  };
}

function attachFactProof(corpus, fact) {
  assert(corpus && corpus.version === 1, "Invalid proof corpus");
  assert(fact && fact.kind === "FACT_PROOF", "Invalid fact proof");
  assert(fact.chain === corpus.chain, "Fact chain mismatch");
  assert(
    corpus.latest_aggregate && fact.corpus_commitment === corpus.latest_aggregate.commitment,
    "Fact must bind to the latest aggregate commitment"
  );

  const factKey = `${fact.slot}:${fact.predicate}:${fact.subject}`;
  return {
    ...corpus,
    fact_commitments: [...corpus.fact_commitments, fact.commitment],
    index: {
      ...corpus.index,
      facts: {
        ...(corpus.index.facts || {}),
        [factKey]: fact.commitment
      }
    }
  };
}

function corpusCommitment(corpus) {
  assert(corpus && corpus.version === 1, "Invalid proof corpus");
  return hashCommitment("PROOF_CORPUS", corpus);
}

module.exports = {
  ETHEREUM_GENESIS_ROOT,
  aggregateSegmentProofs,
  attachAggregate,
  attachFactProof,
  corpusCommitment,
  createFactProof,
  createGenesisAnchor,
  createProofCorpus,
  createSegmentProof,
  hashCommitment,
  verifyContiguousSegments
};
